### PR TITLE
Extend OMiC system with prison and probation details

### PR DESCRIPTION
**Pull request by sldblog**
### Added frequent users on the prison side

  The prison offender managers appear in many interactions, so it seemed to make sense to pull them out.

### Added a direct relationship between NDMIS and OMiC

  The National Application Reporting Team _maintains_ the report, but system sends it (I think)

### Added prison and probation side links

  OMiC is the first service that needs a presence in both workspaces as it pulls information from both places.

  🚨 Consequently, it's not possible to draw its system context diagram without duplicating "unrelated" systems in the other workspace.

## What is the intent behind these changes?

To more closely map the Offender management in custody system to reality.

### CODE REVIEW

File: src/main/kotlin/probation/views.kt [hunk 1]
```diff
@@ -29,7 +29,6 @@ fun probationViews(model: Model, views: ViewSet) {
   val ndmis = model.getSoftwareSystemWithName("NDMIS")!!
   views.createSystemContextView(ndmis, "ndmis-context", null).apply {
     addDefaultElements()
-    addNearestNeighbours(model.getSoftwareSystemWithName("Offender Management Allocation Manager")!!)
```
Comment (sldblog) [inline src/main/kotlin/probation/views.kt:32]: No longer required as the `ndmis.uses(omic.allocationManager, ...)` link will be added in the `addDefaultElements` part

File: src/main/kotlin/probation/model.kt [hunk 2]
```diff
@@ -1,12 +1,16 @@
...
 import com.structurizr.model.Model
 
 import uk.gov.justice.hmpps.architecture.prison.NOMIS
-import uk.gov.justice.hmpps.architecture.prison.OffenderAllocationManager
+import uk.gov.justice.hmpps.architecture.prison.OffenderManagementInCustody
 import uk.gov.justice.hmpps.architecture.shared.Tags
 
 fun probationModel(model: Model) {
+  model.setImpliedRelationshipsStrategy(
+      CreateImpliedRelationshipsUnlessAnyRelationshipExistsStrategy())
```
Comment (spikeheap) [inline src/main/kotlin/probation/model.kt:12]: I'm curious, what does this do?

Comment (sldblog) [reply src/main/kotlin/probation/model.kt:12]: It's a way to automatically add relationships to a system if its container is used.

This line adds a relationship between a system and a container:
```kotlin
ndmis.uses(omic.allocationManager, "sends extracts containing service user allocation to", "email")
```

The strategy above will automatically add a relationship between `ndmis` and `omic.system`.
The goal is to have the connection there for the higher level diagrams.

https://github.com/structurizr/java/blob/v1.4.7/structurizr-core/src/com/structurizr/model/CreateImpliedRelationshipsUnlessAnyRelationshipExistsStrategy.java

File: src/main/kotlin/prison/offenderManagementInCustody.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+    ldu = model.addPerson("Local Divisional Unit").apply {
+      setLocation(Location.Internal)
+    }
+
+    allocationManager = system.addContainer("Offender Management Allocation Manager", "A service for allocating Prisoners to Prisoner Offender Managers (POMs)", "Ruby on Rails").apply {
+      setUrl("https://github.com/ministryofjustice/offender-management-allocation-manager")
+    }
+
+    ldu.uses(allocationManager, "gets notification about ??? from", "gov.uk notify")
```
Comment (sldblog) [inline src/main/kotlin/prison/offenderManagementInCustody.kt:29]: @amitmuthu1 I did not make a note of what they get notified about :( Could you help, please?

**Comment by sldblog** [comment]
♻️ Rebased, now I'll try to apply the pattern introduced in #15

File: src/main/kotlin/prison/model.kt [hunk 4]
```diff
@@ -15,8 +15,15 @@ fun prisonModel(model: Model) {
...
+  val pom = model.addPerson("Prison Offender Manager", "responsible for the service users in their prison")
+  pom.interactsWith(spo, "managed by")
+
   HmmpsAuth(model)
   Delius(model)
   NDH(model).system
   PATHFINDER(model)
+
+  spo.uses(OffenderManagementInCustody.allocationManager, "look at unallocated service users coming from court in")
+  pom.uses(OffenderManagementInCustody.allocationManager, "look at service users who need handing over to community in")
```
Comment (sldblog) [inline src/main/kotlin/prison/model.kt:28]: @spikeheap I wanted to use these within the `defineRelationships`, but that needed
- referencing the people by name, which is not available there (needs a `model` which I didn't want to pass through)
- creating a new holder object for 'internal teams' (which I didn't want to include here as it sounds like a significant thing)

For the time being, I opted to define these person-container relationships here to tackle it later.

File: src/main/kotlin/probation/model.kt [hunk 5]
```diff
@@ -72,9 +77,10 @@ fun probationModel(model: Model) {
   interventionServices.uses(interventionsManager, "creates new interventions in")
   interventionsManager.uses(delius, "pushes contact information of interest to")
   ndmis.uses(delius, "extracts and transforms data from")
-  nomis.uses(delius, "offender data is copied into", "NDH")
-  nomis.uses(oasys, "offender data is coped into", "NDH")
-  nomis.uses(prisonToProbation, "notifies changes")
+  ndmis.uses(OffenderManagementInCustody.allocationManager, "sends extracts containing service user allocation to", "email")
+  NOMIS.system.uses(delius, "offender data is copied into", "NDH")
+  NOMIS.system.uses(oasys, "offender data is coped into", "NDH")
+  NOMIS.system.uses(prisonToProbation, "notifies changes")
```
Comment (spikeheap) [inline src/main/kotlin/probation/model.kt:83]: Hi @sldblog, could these live in `NOMIS.defineRelationships()`? If we're encapsulating that in the `NOMIS` class I think it makes sense to define it's outgoing relationships there.

Comment (sldblog) [reply src/main/kotlin/probation/model.kt:83]: I don't know 😅

The first two systems are probation-specific, which wouldn't exist in `prison/workspace` paths. The last one is a bridging system, so probably worth having it in both.

Comment (sldblog) [reply src/main/kotlin/probation/model.kt:83]: It looks like after #35, yes

**Pull request merged**